README.md
# ethabi

The ABI, Application Binary Interface, is basically how you call functions in a contract and get data back.

> An ABI determines such details as how functions are called and in which binary format information should be passed from one program component to the next...

An Ethereum smart contract is bytecode, EVM, on the Ethereum blockchain. Among the EVM, there could be several functions in a contract. An ABI is necessary so that you can specify which function in the contract to invoke, as well as get a guarantee that the function will return data in the format you are expecting. [read more](http://ethereum.stackexchange.com/a/1171/394)

This library encodes function calls and decodes their output.

[Documentation](https://docs.rs/ethabi)

### Disclaimer

This library intends to support only valid ABIs generated by recent Solidity versions. Specifically, we don't intend to support ABIs that are invalid due to known Solidity bugs or by external libraries that don't strictly follow the specification.
Please make sure to pre-process your ABI in case it's not supported before using it with `ethabi`.

### Installation

  ```
  cargo install ethabi-cli
  ```

### Usage

```
Ethereum ABI coder.
  Copyright 2016-2017 Parity Technologies (UK) Limited

Usage:
    ethabi encode function <abi-path> <function-name-or-signature> [-p <param>]... [-l | --lenient]
    ethabi encode params [-v <type> <param>]... [-l | --lenient]
    ethabi decode function <abi-path> <function-name-or-signature> <data>
    ethabi decode params [-t <type>]... <data>
    ethabi decode log <abi-path> <event-name-or-signature> [-l <topic>]... <data>
    ethabi -h | --help

Options:
    -h, --help         Display this message and exit.
    -l, --lenient      Allow short representation of input params.

Commands:
    encode             Encode ABI call.
    decode             Decode ABI call result.
    function           Load function from json ABI file.
    params             Specify types of input params inline.
    log                Decode event log.
```

### Examples

```
ethabi encode params -v bool 1
```

> 0000000000000000000000000000000000000000000000000000000000000001

--

```
ethabi encode params -v bool 1 -v string gavofyork -v bool 0
```

> 00000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000060000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000096761766f66796f726b0000000000000000000000000000000000000000000000

--

```
ethabi encode params -v bool[] [1,0,false]
```

> 00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000003000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000

--

```
ethabi encode params -v '(string,bool,string)' '(test,1,cyborg)'
```

> 00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000060000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000000000004746573740000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000066379626f72670000000000000000000000000000000000000000000000000000

--

```
ethabi encode function examples/test.json foo -p 1
```

```json
[{
	"type":"function",
	"inputs": [{
		"name":"a",
		"type":"bool"
	}],
	"name":"foo",
	"outputs": []
}]
```

> 455575780000000000000000000000000000000000000000000000000000000000000001

--

```
ethabi encode function examples/test.json foo(bool) -p 1
```

```json
[{
	"type":"function",
	"inputs": [{
		"name":"a",
		"type":"bool"
	}],
	"name":"foo",
	"outputs": []
}]
```

> 455575780000000000000000000000000000000000000000000000000000000000000001

--

```
ethabi encode function examples/test.json bar(bool) -p 1
```

```json
[{
	"type":"function",
	"inputs": [{
		"name":"a",
		"type":"bool"
	}],
	"name":"foo",
	"outputs": []
}]
```

> 6fae94120000000000000000000000000000000000000000000000000000000000000001

--

```
ethabi encode function examples/test.json bar(string):(uint256) -p 1
```

```json
[{
	"type":"function",
	"inputs": [{
		"type":"string"
	}],
	"name":"foo",
	"outputs": [{
		"type": "uint256"
	}]
}]
```

> d473a8ed000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000013100000000000000000000000000000000000000000000000000000000000000

--

```
ethabi decode params -t bool 0000000000000000000000000000000000000000000000000000000000000001
```

> bool true

--

```
ethabi decode params -t bool -t string -t bool 00000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000060000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000096761766f66796f726b0000000000000000000000000000000000000000000000
```

> bool true<br/>
> string gavofyork<br/>
> bool false

--

```
ethabi decode params -t bool[] 00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000003000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
```

> bool[] [true,false,false]

--

```
ethabi decode params -t '(string,bool,string)' 00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000060000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000a0000000000000000000000000000000000000000000000000000000000000000673706972616c000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000067175617361720000000000000000000000000000000000000000000000000000```
```

> (string,bool,string) (spiral,true,quasar)

--

```
ethabi decode function ./examples/foo.json bar 0000000000000000000000000000000000000000000000000000000000000001
```

```json
[{
	"constant":false,
	"inputs":[{
		"name":"hello",
		"type":"address"
	}],
	"name":"bar",
	"outputs":[{
		"name":"",
		"type":"bool"
	}],
	"type":"function"
}]
```

> bool true

--

```
ethabi decode log ./examples/event.json Event -l 0000000000000000000000000000000000000000000000000000000000000001 0000000000000000000000004444444444444444444444444444444444444444
```

> a bool true<br/>
> b address 4444444444444444444444444444444444444444


CHANGELOG.md
# Changelog

The format is based on [Keep a Changelog].

[Keep a Changelog]: http://keepachangelog.com/en/1.0.0/

## [18.0.0] - 2022-11-16
### Added
- Decode function that fails if there is leftover data.

### Changed
- Allow serde in no_std.
- Bump ethereum-types to 0.14.0.
- Updated Rust edition to 2021.

### Fixed
- Nested tuple handling.

### Dependencies
- Bump ethereum-types to 0.14.0

## [17.2.0] - 2022-08-12
### Added
- Add `Token::into_tuple`
- Parse nested tuples/arrays

## [17.1.0] - 2022-06-15
### Added
- Add Serialize trait to Token, Log, LogParams.

### Changed
- Optimize encoder to minimize allocations and copying for ~85% speedup.

## [17.0.0] - 2022-03-01
### Added
- `parity-codec` cargo feature.

### Changed
- Make deprecated function "constant" field optional.
- Allow LenientTokenizer to parse uint256 with units `ether|gwei|nano|nanoether|wei`.
- Bump ethereum-types to 0.13.1.

### Fixed
- Make abi of public library function containing enum parameter parsable.

## [16.0.0] - 2021-12-18
### Added
- Reexport signature functions
- Support solidity error type

### Changed
- Remove anyhow in crates intended to be libraries.

### Fixed
- Fix out of bounds access resulting in panic during decoding of dynamic arrays.

## [15.0.0] - 2021-09-23
### Added
- Optional field internalType to Param.
- Support for no std.
- Method to retrieve the short 4 byte signature of a function.

### Changed
- Bump ethereum-types to 0.12.0.

## [14.1.0] - 2021-07-08
### Added
- `Serialize` support for contracts.

### Fixed
- Fix encoding of nested static tuples in dynamic tuple.
- Fix running out of memory when decoding corrupted array encodings.

## [14.0.0] - 2021-03-31
### Added
- Re-export of ethereum-types.
- Support abiv2 tuples.
- Parse StateMutability in Function abis.
- Support the receive function as an additional operation type.

### Changed
- Update ethereum-types dependency.
- Use lossy UTF-8 decoding for strings.

### Deprecated
- Deprecate `Function::constant`.

### Fixed
- Fix Contract having a receive function by default.
- Fix decoder to handle encoded data with length % 32 != 0.

## [12.0.0]
### Dependencies
- Upgrade ethereum-types (PR #183)

## [11.0.0] - 2020-01-16
### Changed
- Support overloaded contract functions (PR #166)
- Removed `error_chain` and the `backtrace` feature. (#167)
- Update to 2018 edition (PR #171, #172)
- Fix handling of large ints (PR #173)
- Support Tuple parameter types (structs in Solidity) (PR #174)
### Dependencies
- Upgrade syn, proc-macro2, quote and heck crates (PR #169)

## [10.0.0] - 2020-01-08
### Changed
- Replace docopt by structopt (clap) because of performance issue (#161)
### Fixed
- Return an exit code 1 on failure, including wrong input parameters


cli/src/main.rs
use anyhow::anyhow;
use ethabi::{
	decode, encode,
	param_type::{ParamType, Reader},
	token::{LenientTokenizer, StrictTokenizer, Token, Tokenizer},
	Contract, Event, Function, Hash,
};
use itertools::Itertools;
use sha3::{Digest, Keccak256};
use std::fs::File;
use structopt::StructOpt;

#[derive(StructOpt, Debug)]
/// Ethereum ABI coder.
enum Opt {
	/// Encode ABI call.
	Encode(Encode),
	/// Decode ABI call result.
	Decode(Decode),
}

#[derive(StructOpt, Debug)]
enum Encode {
	/// Load function from JSON ABI file.
	Function {
		abi_path: String,
		function_name_or_signature: String,
		#[structopt(short, number_of_values = 1)]
		params: Vec<String>,
		/// Allow short representation of input params.
		#[structopt(short, long)]
		lenient: bool,
	},
	/// Specify types of input params inline.
	Params {
		/// Pairs of types directly followed by params in the form:
		///
		/// -v <type1> <param1> -v <type2> <param2> ...
		#[structopt(short = "v", name = "type-or-param", number_of_values = 2, allow_hyphen_values = true)]
		params: Vec<String>,
		/// Allow short representation of input params (numbers are in decimal form).
		#[structopt(short, long)]
		lenient: bool,
	},
}

#[derive(StructOpt, Debug)]
enum Decode {
	/// Load function from JSON ABI file.
	Function { abi_path: String, function_name_or_signature: String, data: String },
	/// Specify types of input params inline.
	Params {
		#[structopt(short, name = "type", number_of_values = 1)]
		types: Vec<String>,
		data: String,
	},
	/// Decode event log.
	Log {
		abi_path: String,
		event_name_or_signature: String,
		#[structopt(short = "l", name = "topic", number_of_values = 1)]
		topics: Vec<String>,
		data: String,
	},
}

fn main() -> anyhow::Result<()> {
	println!("{}", execute(std::env::args())?);

	Ok(())
}

fn execute<I>(args: I) -> anyhow::Result<String>
where
	I: IntoIterator,
	I::Item: Into<std::ffi::OsString> + Clone,
{
	let opt = Opt::from_iter(args);

	match opt {
		Opt::Encode(Encode::Function { abi_path, function_name_or_signature, params, lenient }) => {
			encode_input(&abi_path, &function_name_or_signature, &params, lenient)
		}
		Opt::Encode(Encode::Params { params, lenient }) => encode_params(&params, lenient),
		Opt::Decode(Decode::Function { abi_path, function_name_or_signature, data }) => {
			decode_call_output(&abi_path, &function_name_or_signature, &data)
		}
		Opt::Decode(Decode::Params { types, data }) => decode_params(&types, &data),
		Opt::Decode(Decode::Log { abi_path, event_name_or_signature, topics, data }) => {
			decode_log(&abi_path, &event_name_or_signature, &topics, &data)
		}
	}
}

fn load_function(path: &str, name_or_signature: &str) -> anyhow::Result<Function> {
	let file = File::open(path)?;
	let contract = Contract::load(file)?;
	let params_start = name_or_signature.find('(');

	match params_start {
		// It's a signature
		Some(params_start) => {
			let name = &name_or_signature[..params_start];

			contract
				.functions_by_name(name)?
				.iter()
				.find(|f| f.signature() == name_or_signature)
				.cloned()
				.ok_or_else(|| anyhow!("invalid function signature `{}`", name_or_signature))
		}

		// It's a name
		None => {
			let functions = contract.functions_by_name(name_or_signature)?;
			match functions.len() {
				0 => unreachable!(),
				1 => Ok(functions[0].clone()),
				_ => Err(anyhow!(
					"More than one function found for name `{}`, try providing the full signature",
					name_or_signature
				)),
			}
		}
	}
}

fn load_event(path: &str, name_or_signature: &str) -> anyhow::Result<Event> {
	let file = File::open(path)?;
	let contract = Contract::load(file)?;
	let params_start = name_or_signature.find('(');

	match params_start {
		// It's a signature.
		Some(params_start) => {
			let name = &name_or_signature[..params_start];
			let signature = hash_signature(name_or_signature);
			contract
				.events_by_name(name)?
				.iter()
				.find(|event| event.signature() == signature)
				.cloned()
				.ok_or_else(|| anyhow!("Invalid signature `{}`", signature))
		}

		// It's a name.
		None => {
			let events = contract.events_by_name(name_or_signature)?;
			match events.len() {
				0 => unreachable!(),
				1 => Ok(events[0].clone()),
				_ => Err(anyhow!(
					"More than one function found for name `{}`, try providing the full signature",
					name_or_signature
				)),
			}
		}
	}
}

fn parse_tokens(params: &[(ParamType, &str)], lenient: bool) -> anyhow::Result<Vec<Token>> {
	params
		.iter()
		.map(|&(ref param, value)| match lenient {
			true => LenientTokenizer::tokenize(param, value),
			false => StrictTokenizer::tokenize(param, value),
		})
		.collect::<Result<_, _>>()
		.map_err(From::from)
}

fn encode_input(path: &str, name_or_signature: &str, values: &[String], lenient: bool) -> anyhow::Result<String> {
	let function = load_function(path, name_or_signature)?;

	let params: Vec<_> =
		function.inputs.iter().map(|param| param.kind.clone()).zip(values.iter().map(|v| v as &str)).collect();

	let tokens = parse_tokens(&params, lenient)?;
	let result = function.encode_input(&tokens)?;

	Ok(hex::encode(result))
}

fn encode_params(params: &[String], lenient: bool) -> anyhow::Result<String> {
	assert_eq!(params.len() % 2, 0);

	let params = params
		.iter()
		.tuples::<(_, _)>()
		.map(|(x, y)| Reader::read(x).map(|z| (z, y.as_str())))
		.collect::<Result<Vec<_>, _>>()?;

	let tokens = parse_tokens(params.as_slice(), lenient)?;
	let result = encode(&tokens);

	Ok(hex::encode(result))
}

fn decode_call_output(path: &str, name_or_signature: &str, data: &str) -> anyhow::Result<String> {
	let function = load_function(path, name_or_signature)?;
	let data: Vec<u8> = hex::decode(data)?;
	let tokens = function.decode_output(&data)?;
	let types = function.outputs;

	assert_eq!(types.len(), tokens.len());

	let result =
		types.iter().zip(tokens.iter()).map(|(ty, to)| format!("{} {to}", ty.kind)).collect::<Vec<String>>().join("\n");

	Ok(result)
}

fn decode_params(types: &[String], data: &str) -> anyhow::Result<String> {
	let types: Vec<ParamType> = types.iter().map(|s| Reader::read(s)).collect::<Result<_, _>>()?;

	let data: Vec<u8> = hex::decode(data)?;

	let tokens = decode(&types, &data)?;

	assert_eq!(types.len(), tokens.len());

	let result =
		types.iter().zip(tokens.iter()).map(|(ty, to)| format!("{ty} {to}")).collect::<Vec<String>>().join("\n");

	Ok(result)
}

fn decode_log(path: &str, name_or_signature: &str, topics: &[String], data: &str) -> anyhow::Result<String> {
	let event = load_event(path, name_or_signature)?;
	let topics: Vec<Hash> = topics.iter().map(|t| t.parse()).collect::<Result<_, _>>()?;
	let data = hex::decode(data)?;
	let decoded = event.parse_log((topics, data).into())?;

	let result = decoded
		.params
		.into_iter()
		.map(|log_param| format!("{} {}", log_param.name, log_param.value))
		.collect::<Vec<String>>()
		.join("\n");

	Ok(result)
}

fn hash_signature(sig: &str) -> Hash {
	Hash::from_slice(Keccak256::digest(sig.replace(' ', "").as_bytes()).as_slice())
}

#[cfg(test)]
mod tests {
	use super::execute;

	#[test]
	fn simple_encode() {
		let command = "ethabi encode params -v bool 1".split(' ');
		let expected = "0000000000000000000000000000000000000000000000000000000000000001";
		assert_eq!(execute(command).unwrap(), expected);
	}

	#[test]
	fn int_encode() {
		let command = "ethabi encode params -v int256 -2 --lenient".split(' ');
		let expected = "fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffe";
		assert_eq!(execute(command).unwrap(), expected);

		let command = "ethabi encode params -v int256 -0 --lenient".split(' ');
		let expected = "0000000000000000000000000000000000000000000000000000000000000000";
		assert_eq!(execute(command).unwrap(), expected);
	}

	#[test]
	fn uint_encode_must_be_positive() {
		let command = "ethabi encode params -v uint256 -2 --lenient".split(' ');
		assert!(execute(command).is_err());
	}

	#[test]
	fn uint_encode_requires_decimal_inputs() {
		let command = "ethabi encode params -v uint256 123abc --lenient".split(' ');
		let result = execute(command);
		assert!(result.is_err());
		let err = result.unwrap_err();
		assert_eq!(err.to_string(), "Uint parse error: InvalidCharacter");
	}

	#[test]
	fn uint_encode_big_numbers() {
		let command =
			"ethabi encode params -v uint256 100000000000000000000000000000000022222222222222221111111111111 --lenient"
				.split(' ');
		let expected = "0000000000003e3aeb4ae1383562f4b82261d96a3f7a5f62ca19599c1ad6d1c7";
		assert_eq!(execute(command).unwrap(), expected);
	}

	#[test]
	fn int_encode_large_negative_numbers() {
		// i256::min_value() is ok
		let command = "ethabi encode params -v int256 -57896044618658097711785492504343953926634992332820282019728792003956564819968 --lenient".split(' ');
		let expected = "8000000000000000000000000000000000000000000000000000000000000000";
		assert_eq!(execute(command).unwrap(), expected);

		// i256::min_value() - 1 is too much
		let command = "ethabi encode params -v int256 -57896044618658097711785492504343953926634992332820282019728792003956564819969 --lenient".split(' ');
		assert_eq!(execute(command).unwrap_err().to_string(), "int256 parse error: Underflow");
	}

	#[test]
	fn int_encode_large_positive_numbers() {
		// Overflow
		let command = "ethabi encode params -v int256 100000000000000000000000000000000022222222222222221111111111111333333333344556 --lenient".split(' ');
		assert_eq!(execute(command).unwrap_err().to_string(), "int256 parse error: Overflow");

		// i256::max_value() is ok
		let command = "ethabi encode params -v int256 57896044618658097711785492504343953926634992332820282019728792003956564819967 --lenient".split(' ');
		let expected = "7fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff";
		assert_eq!(execute(command).unwrap(), expected);

		// i256::max_value() + 1 is too much
		let command = "ethabi encode params -v int256 57896044618658097711785492504343953926634992332820282019728792003956564819968 --lenient".split(' ');
		assert_eq!(execute(command).unwrap_err().to_string(), "int256 parse error: Overflow");
	}

	#[test]
	fn multi_encode() {
		let command = "ethabi encode params -v bool 1 -v string gavofyork -v bool 0".split(' ');
		let expected = "00000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000060000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000096761766f66796f726b0000000000000000000000000000000000000000000000";
		assert_eq!(execute(command).unwrap(), expected);
	}

	#[test]
	fn array_encode() {
		let command = "ethabi encode params -v bool[] [1,0,false]".split(' ');
		let expected = "00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000003000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000";
		assert_eq!(execute(command).unwrap(), expected);
	}

	#[test]
	fn function_encode_by_name() {
		let command = "ethabi encode function ../res/test.abi foo -p 1".split(' ');
		let expected = "455575780000000000000000000000000000000000000000000000000000000000000001";
		assert_eq!(execute(command).unwrap(), expected);
	}

	#[test]
	fn function_encode_by_signature() {
		let command = "ethabi encode function ../res/test.abi foo(bool) -p 1".split(' ');
		let expected = "455575780000000000000000000000000000000000000000000000000000000000000001";
		assert_eq!(execute(command).unwrap(), expected);
	}

	#[test]
	fn nonexistent_function() {
		// This should fail because there is no function called 'nope' in the ABI
		let command = "ethabi encode function ../res/test.abi nope -p 1".split(' ');
		assert!(execute(command).is_err());
	}

	#[test]
	fn overloaded_function_encode_by_name() {
		// This should fail because there are two definitions of `bar in the ABI
		let command = "ethabi encode function ../res/test.abi bar -p 1".split(' ');
		assert!(execute(command).is_err());
	}

	#[test]
	fn overloaded_function_encode_by_first_signature() {
		let command = "ethabi encode function ../res/test.abi bar(bool) -p 1".split(' ');
		let expected = "6fae94120000000000000000000000000000000000000000000000000000000000000001";
		assert_eq!(execute(command).unwrap(), expected);
	}

	#[test]
	fn overloaded_function_encode_by_second_signature() {
		let command = "ethabi encode function ../res/test.abi bar(string):(uint256) -p 1".split(' ');
		let expected = "d473a8ed0000000000000000000000000000000000000000000000000000000000000020\
		                000000000000000000000000000000000000000000000000000000000000000131000000\
		                00000000000000000000000000000000000000000000000000000000";
		assert_eq!(execute(command).unwrap(), expected);
	}

	#[test]
	fn simple_decode() {
		let command =
			"ethabi decode params -t bool 0000000000000000000000000000000000000000000000000000000000000001".split(' ');
		let expected = "bool true";
		assert_eq!(execute(command).unwrap(), expected);
	}

	#[test]
	fn int_decode() {
		let command = "ethabi decode params -t int256 fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffe"
			.split(' ');
		let expected = "int256 fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffe";
		assert_eq!(execute(command).unwrap(), expected);
	}

	#[test]
	fn multi_decode() {
		let command = "ethabi decode params -t bool -t string -t bool 00000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000060000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000096761766f66796f726b0000000000000000000000000000000000000000000000".split(' ');
		let expected = "bool true
string gavofyork
bool false";
		assert_eq!(execute(command).unwrap(), expected);
	}

	#[test]
	fn array_decode() {
		let command = "ethabi decode params -t bool[] 00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000003000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000".split(' ');
		let expected = "bool[] [true,false,false]";
		assert_eq!(execute(command).unwrap(), expected);
	}

	#[test]
	fn abi_decode() {
		let command = "ethabi decode function ../res/foo.abi bar 0000000000000000000000000000000000000000000000000000000000000001".split(' ');
		let expected = "bool true";
		assert_eq!(execute(command).unwrap(), expected);
	}

	#[test]
	fn log_decode() {
		let command = "ethabi decode log ../res/event.abi Event -l 0000000000000000000000000000000000000000000000000000000000000001 0000000000000000000000004444444444444444444444444444444444444444".split(' ');
		let expected = "a true
b 4444444444444444444444444444444444444444";
		assert_eq!(execute(command).unwrap(), expected);
	}

	#[test]
	fn log_decode_signature() {
		let command = "ethabi decode log ../res/event.abi Event(bool,address) -l 0000000000000000000000000000000000000000000000000000000000000001 0000000000000000000000004444444444444444444444444444444444444444".split(' ');
		let expected = "a true
b 4444444444444444444444444444444444444444";
		assert_eq!(execute(command).unwrap(), expected);
	}

	#[test]
	fn nonexistent_event() {
		// This should return an error because no event 'Nope(bool,address)' exists
		let command = "ethabi decode log ../res/event.abi Nope(bool,address) -l 0000000000000000000000000000000000000000000000000000000000000000 0000000000000000000000004444444444444444444444444444444444444444".split(' ');
		assert!(execute(command).is_err());
	}
}
